serde/src/de/value.rs

```diff
@@ -22,31 +23,39 @@ use bytes;
 #[derive(Clone, Debug, PartialEq)]
 pub enum Error {
     /// The value had some syntatic error.
-    SyntaxError,
+    Syntax(String),
+
+    /// The value had an incorrect type.
+    Type(de::Type),
+
+    /// The value had an invalid length.
+    Length(usize),
```

Comment:
The `ValueDeserializer`s are generic in respect to the error now, so the `Length` variant is indeed not used by the existing implementations. This is a remnant of take one to avoid unifying `Length` into `Syntax`, because all nested deserializers are required to go through `value::Error` there which is not the case in take two. In conclusion this variant isn't required by the existing implementations. If we remove it and someone calls `de::Error::length_mismatch` the error would be unified into the `Syntax` variant though, but I guess this is intentional because it's explicitly stated in the `de::Error` trait.